src/ngx_http_lua_socket_tcp.c

```diff
@@ -2793,13 +2806,35 @@ ngx_http_lua_socket_tcp_send(lua_State *L)
 
     switch (type) {
         case LUA_TNUMBER:
+            num = (double) lua_tonumber(L, 2);
+            if (num == (double) (long) num) {
+                b->last = ngx_snprintf(b->last, NGX_INT64_LEN, "%l",
+                                       (long) num);
+
+            } else {
+                /**
+                 * The maximum number of significant digits is 14 in lua.
+                 * Please refer to lj_strfmt.c:411 for more details.
+                 */
+                n = snprintf((char *) b->last, NGX_DOUBLE_LEN, "%.14g", num);
+                if (n < 0) {
+                    ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, ngx_errno,
+                                  "snprintf(\"%f\") failed");
+
+                } else {
+                    b->last += n;
+                }
+            }
```

Comment:
Couldn't we avoid so much code duplication by making the len calculation and the buffer write into two inlined functions instead?